native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeBuildMojo.java

```diff
@@ -41,114 +41,21 @@
-        }
-
-        evaluator = new PluginParameterExpressionEvaluator(session, mojoExecution);
-        maybeSetMainClassFromPlugin(this::consumeExecutionsNodeValue, "org.apache.maven.plugins:maven-shade-plugin", "transformers", "transformer", "mainClass");
-        maybeSetMainClassFromPlugin(this::consumeConfigurationNodeValue, "org.apache.maven.plugins:maven-assembly-plugin", "archive", "manifest", "mainClass");
-        maybeSetMainClassFromPlugin(this::consumeConfigurationNodeValue, "org.apache.maven.plugins:maven-jar-plugin", "archive", "manifest", "mainClass");
-        maybeAddGeneratedResourcesConfig(buildArgs);
-        buildImage();
-    }
-
-    private String consumeConfigurationNodeValue(String pluginKey, String... nodeNames) {
-        Plugin selectedPlugin = project.getPlugin(pluginKey);
-        if (selectedPlugin == null) {
-            return null;
-        }
-        return getConfigurationNodeValue(selectedPlugin, nodeNames);
-    }
-
-    private String consumeExecutionsNodeValue(String pluginKey, String... nodeNames) {
-        Plugin selectedPlugin = project.getPlugin(pluginKey);
-        if (selectedPlugin == null) {
-            return null;
-        }
-        for (PluginExecution execution : selectedPlugin.getExecutions()) {
-            String value = getConfigurationNodeValue(execution, nodeNames);
-            if (value != null) {
-                return value;
-            }
-        }
-        return null;
-    }
-
-    private String getConfigurationNodeValue(ConfigurationContainer container, String... nodeNames) {
-        if (container != null && container.getConfiguration() instanceof Xpp3Dom) {
-            Xpp3Dom node = (Xpp3Dom) container.getConfiguration();
-            for (String nodeName : nodeNames) {
-                node = node.getChild(nodeName);
-                if (node == null) {
-                    return null;
-                }
-            }
-            String value = node.getValue();
-            return evaluateValue(value);
-        }
-        return null;
-    }
-
-    private String evaluateValue(String value) {
-        if (value != null) {
-            try {
-                Object evaluatedValue = evaluator.evaluate(value);
-                if (evaluatedValue instanceof String) {
-                    return (String) evaluatedValue;
-                }
-            } catch (ExpressionEvaluationException ignored) {
-            }
-        }
-
-        return null;
-    }
-
-    private void maybeSetMainClassFromPlugin(BiFunction<String, String[], String> mainClassProvider, String pluginName, String... nodeNames) {
-        if (mainClass == null) {
-            mainClass = mainClassProvider.apply(pluginName, nodeNames);
-
-            if (mainClass != null) {
-                getLog().info("Obtained main class from plugin " + pluginName + " with the following path: " + String.join(" -> ", nodeNames));
-            }
-        }
-    }
+@Execute(phase = LifecyclePhase.COMPILE)
```

Comment:
@snicoll This works in my testing but I'm not sure if this is a correct thing to do.
Are there cases where resources / additional code is added after the COMPILE phase?

I didn't want to execute PACKAGE phase here since I cannot prevent it from building native-image twice.

Reply:
Compile won't work as it wouldn't have produced the `jar` that NBT requires by default. It should be `package`. 

> I didn't want to execute PACKAGE phase here since I cannot prevent it from building native-image twice.

That's exactly what the `no-fork` variant is for.